### PR TITLE
Use a more efficient way to obtain leader epoch in RemoteLogManager#read

### LINKED ISSUE #30: Use a more efficient way to obtain leader epoch in RemoteLogManager#read
Use a more efficient way to obtain leader epoch in RemoteLogManager#read

### PR DESCRIPTION
**Pull request by divijvaidya**
This is a backport of the changes in obtaining leader epochs from trunk [PR](https://github.com/apache/kafka/pull/11390).

Resolves https://github.com/satishd/kafka/issues/30

### CODE REVIEW

**Comment by divijvaidya** [comment]
@satishd @wyuka @kamalcph Please take a look.

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala [hunk 1]
```diff
@@ -612,23 +616,17 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
...
-    fetchLog(tp).foreach { log =>
-      log.leaderEpochCache.foreach(cache => {
-        var epoch = cache.latestEpoch
-        while (!rlsMetadata.isPresent && epoch.isDefined) {
-          rlsMetadata = fetchRemoteLogSegmentMetadata(tp, epoch.get, offset)
-          epoch = cache.findPreviousEpoch(epoch.get)
-        }
-      })
+    val epoch = fetchLog(tp).flatMap(log => log.leaderEpochCache.flatMap(cache => cache.epochForOffset(offset)))
+    val rlsMetadata = if (epoch.isDefined) {
```
Comment (wyuka) [inline core/src/main/scala/kafka/log/remote/RemoteLogManager.scala]: Nit: In the spirit of keeping the code functional and Scala-like, should we do
```
val rlsMetadata = epoch.flatMap(epoch => fetchRemoteLogSegmentMetadata(tp, epoch, offset))
```

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala [hunk 2]
```diff
@@ -612,23 +616,17 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
...
+      fetchRemoteLogSegmentMetadata(tp, epoch.get, offset)
+    } else {
+      Optional.empty()
     }
 
     if (!rlsMetadata.isPresent) {
+      val epochStr = if (epoch.isDefined) epoch.get.toString else "NOT AVAILABLE"
```
Comment (wyuka) [inline core/src/main/scala/kafka/log/remote/RemoteLogManager.scala]: Nit (I'm really not sure if this is better though):

```
val epochStr = epoch.map(epoch => epoch.toString).getOrElse("NOT AVAILABLE")
```

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala (hunk 1 above)
Comment (kamalcph) [reply core/src/main/scala/kafka/log/remote/RemoteLogManager.scala]: +1

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala (hunk 2 above)
Comment (kamalcph) [reply core/src/main/scala/kafka/log/remote/RemoteLogManager.scala]: +1

**Review by kamalcph — APPROVED** [review]
Thanks for fixing this! LGTM

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala (hunk 2 above)
Comment (divijvaidya) [reply core/src/main/scala/kafka/log/remote/RemoteLogManager.scala]: Thanks for the suggestion. Using if/else with `Option` in scala is considered as anti-practice (aka anti-functional programming), hence, your suggestion completely makes sense. Fixed.

File: core/src/main/scala/kafka/log/remote/RemoteLogManager.scala (hunk 1 above)
Comment (divijvaidya) [reply core/src/main/scala/kafka/log/remote/RemoteLogManager.scala]: Thank you for the suggestion. Fixed.

**Comment by divijvaidya** [comment]
@wyuka @kamalcph, I have addressed your comments. This is ready for review again.

**Comment by satishd** [comment]
Thanks @divijvaidya for addressing the review comments.

**Review by satishd — APPROVED** [review]

**Pull request merged**